File: scripts/pos_core/pix_brl.py
#!/usr/bin/env python3
"""
ZeroClaw Solana POS Agent - Brazil EMV QRCPS & PIX QR Generator Core Module
"""

def calculate_pix_crc16(payload_without_crc: str) -> str:
    """
    Calculates EMV QRCPS CRC16 (CCITT-FALSE, polynomial 0x1021, init 0xFFFF).
    Appends '6304' before computing checksum as per EMV Co / BR Code specification.
    Returns 4-character uppercase hexadecimal string (e.g. '1D2C').
    """
    data_to_hash = (payload_without_crc + "6304").encode('utf-8')
    crc = 0xFFFF
    for byte in data_to_hash:
        crc ^= (byte << 8)
        for _ in range(8):
            if crc & 0x8000:
                crc = ((crc << 1) ^ 0x1021) & 0xFFFF
            else:
                crc = (crc << 1) & 0xFFFF
    return f"{crc:04X}"

def generate_pix_emv_payload(pix_key: str, amount_brl: float, merchant_name: str = "ZeroClaw POS") -> str:
    """
    Generates Brazil-first EMV QRCPS PIX payload with valid CRC16 CCITT-FALSE checksum.
    Compatible with Brazilian banking apps (br.gov.bcb.pix).
    Guarantees Tag 59 byte length <= 99 bytes to prevent EMV parsing failures.
    """
    amount_str = f"{amount_brl:.2f}"
    merchant_name = merchant_name or "ZeroClaw POS"
    pix_key = pix_key or ""
    merchant_bytes = merchant_name.encode('utf-8')

    if len(merchant_bytes) > 99:
        merchant_bytes = merchant_bytes[:99]
        merchant_name = merchant_bytes.decode('utf-8', errors='ignore')
        merchant_bytes = merchant_name.encode('utf-8')
        
    merchant_len = len(merchant_bytes)
    pix_key_bytes = pix_key.encode('utf-8')
    pix_key_len = len(pix_key_bytes)
    payload_base = (
        "00020126580014br.gov.bcb.pix"
        f"01{pix_key_len:02d}{pix_key}"
        "520400005303986"
        f"54{len(amount_str):02d}{amount_str}"
        "5802BR"
        f"59{merchant_len:02d}{merchant_name}"
        "6009SAO PAULO"
        "62070503***"
    )
    crc_hex = calculate_pix_crc16(payload_base)
    return f"{payload_base}6304{crc_hex}"

File: pos-backend/src/domain/pix_brl.rs
/// Brazil EMV QRCPS PIX QR Code Generator
/// Calculates EMV QRCPS CRC16 (CCITT-FALSE, polynomial 0x1021, init 0xFFFF).
pub fn calculate_pix_crc16(payload_without_crc: &str) -> String {
    let data_to_hash = format!("{}6304", payload_without_crc);
    let bytes = data_to_hash.as_bytes();

    let mut crc: u16 = 0xFFFF;
    for &byte in bytes {
        crc ^= (byte as u16) << 8;
        for _ in 0..8 {
            if crc & 0x8000 != 0 {
                crc = (crc << 1) ^ 0x1021;
            } else {
                crc <<= 1;
            }
            crc &= 0xFFFF;
        }
    }

    format!("{:04X}", crc)
}

/// Generates Brazil EMV QRCPS PIX payload with valid CRC16 checksum.
pub fn generate_pix_emv_payload(pix_key: &str, amount_brl: f64, merchant_name: &str) -> String {
    generate_pix_emv_payload_with_txid(pix_key, amount_brl, merchant_name, "101")
}

/// Generates Brazil EMV QRCPS PIX payload with dynamic invoice TxID reference and recalculates Tag 62 length.
pub fn generate_pix_emv_payload_with_txid(
    pix_key: &str,
    amount_brl: f64,
    merchant_name: &str,
    invoice_id: &str,
) -> String {
    let amount_str = format!("{:.2}", amount_brl);
    let merchant_name = if merchant_name.is_empty() {
        "ZeroClaw POS"
    } else {
        merchant_name
    };

    // Truncate merchant name to 99 chars (UTF-8 safe)
    let merchant_truncated: String = merchant_name.chars().take(99).collect();
    // Truncate pix key to 99 chars (UTF-8 safe)
    let pix_key_truncated: String = pix_key.chars().take(99).collect();

    let merchant_len = merchant_truncated.len();
    let pix_key_len = pix_key_truncated.len();

    // Construct Field 05 (TxID reference) inside Tag 62
    let raw_txid = if invoice_id.is_empty() {
        "101".to_string()
    } else {
        format!(
            "INV{}",
            invoice_id
                .trim_start_matches("INV-")
                .trim_start_matches('#')
        )
    };
    let txid_clean: String = raw_txid
        .chars()
        .filter(|c| c.is_ascii_alphanumeric())
        .take(25)
        .collect();

    let field05 = format!("05{:02}{}", txid_clean.len(), txid_clean);
    let tag62 = format!("62{:02}{}", field05.len(), field05);

    let payload_base = format!(
        "00020126580014br.gov.bcb.pix\
         01{:02}{}\
         520400005303986\
         54{:02}{}\
         5802BR\
         59{:02}{}\
         6009SAO PAULO\
         {}",
        pix_key_len,
        pix_key_truncated,
        amount_str.len(),
        amount_str,
        merchant_len,
        merchant_truncated,
        tag62
    );

    let crc_hex = calculate_pix_crc16(&payload_base);
    format!("{}6304{}", payload_base, crc_hex)
}
